[libcxx] adds ranges::fold_left_with_iter and ranges::fold_left

cjdb opened the pull request:
Notable things in this commit does:

* refactors `__indirect_binary_left_foldable`, making it slightly different (but equivalent) to _`indirect-binary-left-foldable`_, which improves readability (a [patch to the Working Paper][patch] was made)
* omits `__cpo` namespace, since it is not required for implementing niebloids (a cleanup should happen in 2024)
* puts tests ensuring invocable robustness and dangling correctness inside the correctness testing to ensure that the algorithms' results are still correct

[patch]: https://github.com/cplusplus/draft/pull/6734

cjdb committed:
[libcxx] extends test range types to take a value-type

This increases the types' versatility so that they're not restricted
just to `int*`.

cjdb committed:
[libcxx] adds `ranges::fold_left_with_iter` and `ranges::fold_left`

cjdb committed:
changes to using static opertaor()

hawkinsw commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+
+  template <input_range _Rp, class _Tp, __indirectly_binary_left_foldable<_Tp, iterator_t<_Rp>> _Fp>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto operator()(_Rp&& __r, _Tp __init, _Fp __f) {
+    auto __result = operator()(ranges::begin(__r), ranges::end(__r), std::move(__init), std::ref(__f));
```
It looks like there is a slight difference between the implementation here and the spec:
```C++
ranges::fold_left_with_iter(std::move(first), last, std::move(init), f).value
```
from [[alg.fold.1]](https://eel.is/c++draft/alg.fold#1). I am not nearly enough of an expert to know if you are doing something special here or if it's just a mistake. 

I hope that this is helpful.

cjdb replied:
The section in [alg.fold] that you've highlighted is for `fold_left`, but this is the range overload for `fold_left_with_iter`. It's organised in a bit of a confusing way (and I plan to reorganise it because I got lost about a dozen times while implementing this). Is that the reason for the discrepancy, or were you intending to apply this to `fold_left` below?

hawkinsw replied:
Sorry! I clearly meant to apply this below! I totally goofed. Sorry. As I look down further, the same `std::move` seems to be present. I will move this comment there. Sorry!

cjdb committed:
adds missing export

cjdb committed:
changes requirements test from asserts to verify

hawkinsw commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+  template <input_iterator _Ip, sentinel_for<_Ip> _Sp, class _Tp, __indirectly_binary_left_foldable<_Tp, _Ip> _Fp>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto
+  operator()(_Ip __first, _Sp __last, _Tp __init, _Fp __f) {
+    return fold_left_with_iter(std::move(__first), std::move(__last), std::move(__init), std::ref(__f)).result;
```
**Moving from above ...**

It looks like there is a slight difference between the implementation here and the spec:
```C++
ranges::fold_left_with_iter(std::move(first), last, std::move(init), f).value
```
from [[alg.fold.1]](https://eel.is/c++draft/alg.fold#1). I am not nearly enough of an expert to know if you are doing something special here or if it's just a mistake. 

I hope that this is helpful.

cjdb replied:
Since [alg.fold]/p1 is a _Returns_ element, we just need to ensure that what we return matches what the standard says it should return. 

Both differences minimise the number of unnecessary copies (which might be expensive). Had the paragraph been "_Effects:_ Equivalent to" instead, then we'd need to be more cautious and ensure that `last` and `f` are copied.

hawkinsw replied:
Thank you!!

cjdb committed:
adds missing file

EricWF commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+  }
+};
+
+inline namespace __cpo {
```
Why the `__cpo` namespace?

I know it's my ignorance, but could you explain the purpose?

cjdb replied:
In this case, it's not required, but it's the way libc++ does things right now. None of the algorithms should be declared in an inline namespace, but I think that's a bigger clean-up. We should decide whether new algorithms get put in this namespace though.

EricWF commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+};
+
+inline namespace __cpo {
+inline constexpr auto fold_left = __fold_left();
```
Why inline?

cjdb replied:
`constexpr` doesn't imply `inline` for objects :(

EricWF commented on libcxx/test/libcxx/diagnostics/ranges.nodiscard_extensions.verify.cpp:
```diff
@@ -87,4 +89,15 @@ void test() {
+
+#if TEST_STD_VER >= 23
+  std::ranges::fold_left(range, 0, std::plus());
+  // expected-warning@-1{{ignoring return value of function declared with 'nodiscard' attribute}}
```
I don't think this is correct. There are many valid ways to use this function without inspecting the result.

cjdb replied:
Can you provide an example please? I'm not aware of any.

philnik777 replied:
If you agree that it's not good code, I don't really see the problem with warning here. There's lots of correct code that the compiler generates warnings on. You wouldn't want to remove `[[nodiscard]]` from `std::count` just because there are valid ways to use the function without inspecting the return value, would you?

cjdb replied:
I think it's substantially better to be explicit about wanting to discard the result, since the primary function of this algorithm is to return a computed result. I also expect the common case to be to want the result, rather than not.

```cpp
(void)std::ranges::fold_left(r, 0, side_effect_f); // clearly discarding the result
```

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/requirements.compile.pass.cpp:
```diff
@@ -0,0 +1,99 @@
+#include <ranges>
+
+#include "test_iterators.h"
+#include "../requirements.h"
```
No relative includes please.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/return_types.compile.pass.cpp:
```diff
@@ -0,0 +1,171 @@
+
+namespace common_ranges {
+template <class T>
+using forward_range = test_common_range<forward_iterator, T>;
```
Rather than spending a bunch of time and effort on tests that only check compile time properties,
consider writing tests that actually run using the same types. It verifies all these tests do and more.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/requirements.verify.cpp:
```diff
@@ -0,0 +1,203 @@
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+// Checks that `std::ranges::fold_left_with_iter`'s requirements are correct.
```
This comment doesn't say much.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/requirements.verify.cpp:
```diff
@@ -0,0 +1,203 @@
+
+// expected-error@*:* 19 {{no matching function for call to object of type 'const __fold_left_with_iter'}}
+
+void test_iterator() {
```
This test is very fragile and will break anytime even the slightest bit of the diagnostic changes.

Also the notes are so loosely coupled to source lines or instantiations it's unclear if they actually attach to the code they're next to.

Consider deleting this test entirely.

cjdb replied:
These were originally static_asserts, but you'd mentioned that you didn't like those, and that a verify test would be better. We need a way to ensure that the requirements are present, and this is the only alternative.

EricWF commented on libcxx/test/std/iterators/iterator.requirements/iterator.concepts/iterator.concept.input/input_iterator.compile.pass.cpp:
```diff
@@ -81,6 +81,7 @@ struct not_indirectly_readable {
 };
 static_assert(!std::indirectly_readable<not_indirectly_readable> && !std::input_iterator<not_indirectly_readable>);
 
+// FIXME(cjdb): deduplicate
```
Or leave duplicated. Duplication in tests is OK.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/requirements.h:
```diff
@@ -0,0 +1,84 @@
+  non_movable(non_movable&&) = delete;
+};
+
+struct copyable_non_movable {
```
That type is not copyable.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
"valid.pass.cpp" could use a different name.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+  }
+  {
+    auto data   = std::vector{-1.1, -2.2, -3.3, -4.4, -5.5, -6.6};
+    auto plus   = [](int const x, double const y) { return x + y; };
```
How many of these tests would fail if we were to swap the implementation with `fold_right`?

cjdb replied:
Good question. Anything that takes a floating-point value is at risk, but this test is only for left folds.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+    auto plus   = [](int const x, double const y) { return x + y; };
+    auto result = std::ranges::fold_left_with_iter(data.begin(), data.end(), 0.0, plus);
+
+    assert(result.in == data.end());
```
Does this hold when I turn on optimizations?

cjdb replied:
I don't understand this question.

EricWF replied:
We're doing equality on floating point numbers. That's sensitive to optimizations.

cjdb replied:
Provided CI is checking at `-On` and is green, yes.

cjdb committed:
repeats for `fold_left`

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+#include "test_range.h"
+#include "../gaussian_sum.h"
+
+constexpr bool test() {
```
I would add runtime tests for:

1. An empty range
2. A range where summing in any order other than `fold_left` would result in an overflow or other invalid condition.
3. That we don't copy the functor like we intend.
4. That the functor is applied the correct amount of times.
5. On non-integral or builtin types.
6. On types other than vector & its iterators.

I understand the signatures for a lot of these configurations are checked via meta-programming elsewhere, but we need to run code to test them too. What if there's a lifetime bug that wouldn't be caught unless the type were' folding allocates?

cjdb replied:
* (1) is already done (lines 29--41)
* (2) and (5) can be coupled together using `std::basic_string<some-small-allocator>`.
* I'm not sure I understand what you're asking for in (3). I've already got cases where the function object is copied and when it's directly initialised.
* Thanks for catching (4)
* I'm not really sure we'll get more value out of (6) than `return_types.compile.pass.cpp`, which checks this.

EricWF replied:
* Please add more tests for (1). What if we're dereferencing the first iterator? In all those cases that's valid because it's actually backed by an element.

(3) What's the purpose of wrapping the binary functor in `std::ref`? To prevent it from being copied repeatedly in case it has state. It would be a meaningful regression were we to start copying the binary operator more.

(6) Almost all of the tests are over contiguous ranges of densely packed objects. So code using more opaque or non-contiguous iterators will compile entirely differently, have wildly different memory access patterns, and could tickle all sorts of bugs we don't know we have.

cjdb replied:
(1) Fair enough.
(3) Ah, good idea.
(6) I still don't think that's going to expose any issues in `fold_left`. I'll change to different types (it's hardly any effort), but still I'm not convinced we're going to get broader runtime coverage.

cjdb replied:
Completed.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+    auto result = std::ranges::fold_left_with_iter(data.begin(), data.end(), 0, std::plus());
+
+    assert(result.in == data.end());
+    assert(result.result == gaussian_sum(data)); // sum of n ascending odd numbers = n^2
```
Just a curiosity, what's the purpose of using gaussian_sum here?

cjdb replied:
It ensures that this test is correct upon inspection. However, it needs to be renamed: https://en.wikipedia.org/wiki/Gauss_sum#:~:text=Not%20to%20be%20confused%20with%20Triangular%20Number%20%C2%A7%C2%A0Formula.

EricWF replied:
The Gaussian sum tests are **brilliant!**
I'm so in love with them.
I had to do a ton of work to get there, bit I'm glad I did. There's an elegance and beauty usings maths to avoid the trap of computing the expected result the same way twice, both incorrectly.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/return_types.compile.pass.cpp:
```diff
@@ -0,0 +1,171 @@
+template <class T>
+using forward_range = test_range<forward_iterator, T>;
+
+static_assert(requires(forward_range<int> r) {
```
Maybe some tests with short that ensure it doesn't promote to int if we don't intend it to?

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+    auto plus         = [](int const x, double const y) { return x + y; };
+    auto const result = std::ranges::fold_left(data.begin(), data.end(), 0.0, plus);
+
+    assert(result == -21.6); // int(  0.0) + -1.1 =   0 + -1.1 =  -1.1
```
I really like the comment here. Nice.

EricWF reviewed (commented):
I hear an updated change is coming, so I'm submitting my comments.

cjdb commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+#if _LIBCPP_STD_VER >= 23
+
+namespace ranges {
+template <class _Ip, class _Tp>
```
clang-format is happy.

cjdb commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+
+template <class _Fp, class _Tp, class _Ip, class _Rp, class _Up = decay_t<_Rp>>
+concept __indirectly_binary_left_foldable_impl =
+    convertible_to<_Rp, _Up> &&                    //
```
This `convertible_to` is at the [bottom of _`indirectly-binary-left-foldable`_](https://eel.is/c++draft/algorithm.syn#concept:indirectly-binary-left-foldable:~:text=convertible_to%3Cinvoke_result_t%3CF%26%2C%20T%2C%20iter_reference_t%3CI%3E%3E%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20decay_t%3Cinvoke_result_t%3CF%26%2C%20T%2C%20iter_reference_t%3CI%3E%3E%3E%3E). I moved it into `__indirectly_binary_left_foldable_impl` to simplify the concept so we don't have to spell out `invoke_result_t<...>` more than once, and [asked it to be updated in the synopsis too](https://github.com/cplusplus/draft/pull/6734). The position of this concept's evaluation hasn't changed.

EricWF replied:
For the room: Chris and I spent 20 minute discussing this offline.

OK, I don't see any value in the refactoring here. Practically speaking it gets us nothing.

However,, writing it out verbatim makes the code trivial to inspect for correctness. You simply diff it against the standard. 

As written this code is significantly harder to understand in the light of the subtleties of concept subsumption.

If I find time I'll be writing some docs for review that suggest we _always_ copy concepts given as exposition verbatim from the standard. Isn't that the whole reason they're there. 

I won't block this review on it. So I'm going to leave this conversation here so we can have it in a more appropriate venue.

hawkinsw replied:
FWIW (and it's probably not much), as I was reviewing this PR, the difference between Chris' awesome implementation and the way that it was written in the standard did force me to spend a significant amount of time determining that Chris was following the standard correctly but his implementation was just smarter than I could comprehend.

In other words, I agree that it would be valuable (for reviewers, if nothing else) for these concepts to be carried verbatim from the spec, if possible.

var-const replied:
FWIW, I'm very much in favor of copying verbatim from the Standard unless there's a strong reason not to.

cjdb replied:
I agree that we need to be copying verbatim for user-facing concepts. Exposition-only concepts are a different story because:

* they're only present for the purposes of conveying the necessary constraints without needing to open a _Constraints_ element;
* are optimised for communicating information in standardese, not for real-world software engineering;
* can be refactored editorially, meaning that they can silently change and we'd be out of sync without realising

Had this been a _Constraints_ element instead of a concept, we would've needed to construct the concept from prose, and it would look as it does here. I think it would be in our users' best interests to treat exposition-only concepts as a shorthand for a _Constraints_ element instead of something that's set in stone. In this case, it only mildly improves readability, and so it's not really worth debating, but there are other cases where it can actually matter (such as _`indirectly-binary-right-foldable`_, but we can have that discussion when I implement that).

EricWF replied:
Can you please specify in some way that this is _not_ copied verbatim? That way nobody has to find this conversation to figure out if it was intentional?

cjdb replied:
Done.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/gaussian_sum.h:
```diff
@@ -0,0 +1,19 @@
+#include <vector>
+
+template <class T>
+constexpr auto gaussian_sum(std::vector<T> const& input) {
```
I'm a bit uncomfortable by the amount of assumptions this function makes about its inputs in order to produce correct output.

cjdb replied:
Fixed.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/requirements.h:
```diff
@@ -0,0 +1,84 @@
+  friend int operator+(copyable_non_movable const&, non_movable const&);
+  friend int operator+(non_movable const&, copyable_non_movable const&);
+
+  friend copyable_non_movable const& operator-(int, copyable_non_movable const&);
```
What is this type? 

Are the weird operator overloads  tied to the copyable_non_movable-ness? They seem like entirely separate concerns.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/requirements.h:
```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
Could you please take another pass at naming these types. They often name concepts they don't fulfill or fulfill only in one particular case. 

Consider moving some of them closer to their test.

cjdb replied:
PTAL

EricWF reviewed (changes requested):
My only blocking concern is the clang verify tests. They're too brittle and will break. 
Otherwise, This change LGTM and is ready to land afterwards.

tcanens commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+  template <input_iterator _Ip, sentinel_for<_Ip> _Sp, class _Tp, __indirectly_binary_left_foldable<_Tp, _Ip> _Fp>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto
+  operator()(_Ip __first, _Sp __last, _Tp __init, _Fp __f) {
+    return fold_left_with_iter(std::move(__first), std::move(__last), std::move(__init), std::ref(__f)).result;
```
There are two moves here (one in `fold_left_with_iter` to pack `__result` into the returned struct, and another one here to unpack it that could in principle be elided due to the _Returns:_ element. (Eliding both would depend on NRVO kicking in.)

EricWF reviewed (approved):
> My only blocking concern is the clang verify tests. They're too brittle and will break. Otherwise, This change LGTM and is ready to land afterwards.

OK, so Chris and I discussed it, and we want to resolve this by reverting to these type of tests:
https://github.com/llvm/llvm-project/pull/75259/commits/bec1572ab327212aaf57af020678403cf7b1a321

So LGTM once that has been addressed! Let's land this!

cjdb committed:
Revert "changes requirements test from asserts to verify"

This reverts commit bec1572ab327212aaf57af020678403cf7b1a321.

cjdb committed:
Revert "adds missing file"

This reverts commit 6a57523fa1d33f2e8d221858c38af8765396c1aa.

cjdb committed:
Revert "repeats for `fold_left`"

This reverts commit 12f36c42514b24a39c02c25bf33c94cf8bef9bc9.

var-const commented on libcxx/test/std/library/description/conventions/customization.point.object/niebloid.compile.pass.cpp:
```diff
@@ -83,6 +83,10 @@ static_assert(test(std::ranges::find_end, a, a));
 static_assert(test(std::ranges::find_first_of, a, a));
```
Can you also take a look at various `robust` tests and add the new algorithms where it makes sense (e.g. `libcxx/test/std/algorithms/ranges_robust_against_omitting_invoke.pass.cpp` looks relevant)?

cjdb replied:
I'm a bit hesitant to use that test file, so `left_folds.pass.cpp` does checks to ensure invocable robustness. I'll my concerns in the new year, when I tackle the other folds.

var-const reviewed (changes requested):
Thanks a lot for working on this!

cjdb committed:
consolidates requirements.compile.pass.cpp for the fold algos

Per @ericwf's feedback, I've consolidated all the requirements tests
into a single file in order to eliminate duplication among the various
fold algorithms (which have similar requirements).

cjdb committed:
eliminates unnecessary `__cpo` namespace

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+#include "../gaussian_sum.h"
+
+constexpr bool test() {
+  {
```
Can you add a brief comment to each test case to explain what it tests? It can be really short -- it just makes it easier to see what's being tested at a glance.

cjdb replied:
I've instead put each test into its own test case function.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    auto range = std::span(data.data(), 0);
+    assert(std::ranges::fold_left(range, 0, std::plus()) == result);
+  }
```
Nit: please separate different test cases by blank lines.

cjdb replied:
I've partially done this. There are some cases where the test cases are identical, but have extra things attached, and having those be adjacent is sorta good for showing that? WDYT?

cjdb replied:
I've completely separated them per more recent feedback below.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+}
+
+int main() {
+  test();
```
We normally test range algorithms with different kinds of iterators and sentinels, can you please templatize `test` and try different types from `test_iterators.h`?

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+// template<input_iterator I, sentinel_for<I> S, class T,
```
We should also test that the algorithm SFINAEs away when given types not satisfying the constraints, unless you think that's redundant?

cjdb replied:
Resolving, as that's handled by the (incoming) `requirements.compile.pass.cpp`.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
```
Nit: please sort the includes.

cjdb replied:
This is a clang-format issue.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/valid.pass.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    assert(result == 0);
+
+    auto range = std::span(data.data(), 0);
```
Nit: missing `<span>` include.

cjdb commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace ranges {
+template <class _Ip, class _Tp>
+struct in_value_result {
+  _Ip in;
+  _Tp result;
+};
```
This is not-to-spec.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+  return true;
+}
+
+int main() {
```
Please add the parameters to `main` (we have some platforms that don't support `main()` with no arguments, weird as that sounds).

cjdb replied:
Oh, I forgot about that.

var-const commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left_with_iter/valid.pass.cpp:
```diff
@@ -0,0 +1,116 @@
+    assert(rresult.result == result.result);
+  }
+  {
+    auto data         = std::vector{2, 4, 6, 8, 10, 12};
```
Nit: the other test file doesn't use CTAD for initializing vectors. I'm ok with either, but can this be consistent?

cjdb replied:
I'm wondering if we should consolidate `fold_left` and `fold_left_with_iter` into a single file, which would resolve this a different way.

var-const reviewed (changes requested):
Finished the first pass. This looks pretty good, just a few comments.

philnik777 commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/requirements.verify.cpp:
```diff
@@ -0,0 +1,202 @@
+  // expected-note@*:*  10 {{candidate template ignored: constraints not satisfied}}
+  // expected-note@*:*  10 {{candidate function template not viable: requires 3 arguments, but 4 were provided}}
+
+  std::ranges::fold_left(bad_iterator_category(), std::unreachable_sentinel, 0, std::plus());
```
It looks like this test checks SFINAEing bahaviour. If I'm not missing something these should just be test to make sure the functions SFINAE away and not something this specific to the implementation. This looks to me like it's testing the compiler and not the library.

cjdb replied:
Resolving so that there's one place for discussion to happen (a reply above).

philnik777 commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/return_types.compile.pass.cpp:
```diff
@@ -0,0 +1,171 @@
+concept is_T = std::same_as<Result, T>;
+
+using std::ranges::fold_left;
+[[maybe_unused]] auto f = [](int x, double y) { return x * y; };
```
Why is this `[[maybe_unused]]``?

cjdb replied:
Clang gets unhappy when I remove it. I think it might be a bug, but I haven't had time to investigate it.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/requirements.verify.cpp:
```diff
@@ -0,0 +1,202 @@
+// expected-error@*:* 19 {{no matching function for call to object of type 'const __fold_left'}}
+
+void test_iterator() {
+  // expected-note@*:*  10 {{candidate template ignored: constraints not satisfied}}
```
Per a video call with @EricWF, I'm going to revert https://github.com/llvm/llvm-project/commit/bec1572ab327212aaf57af020678403cf7b1a321 and the two commits that come after it, which will probably address what you and @philnik777 have expressed concerns over.

cjdb committed:
reworks valid.cpp to incorporate feedback

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/requirements.verify.cpp:
```diff
@@ -0,0 +1,202 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Completed.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/fold_left/requirements.verify.cpp:
```diff
@@ -0,0 +1,202 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
```
> (it presumes implementation details that are only true of libc++)

Which details are those?

cjdb replied:
Ah, this comment was made before my revert, so it's no longer applicable.

EricWF commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -89,9 +89,7 @@ struct __fold_left_with_iter {
 };
 
-inline namespace __cpo {
 inline constexpr auto fold_left_with_iter = __fold_left_with_iter();
```
Please make sure that in your commit message for this change (so it appears in blame easily), you clearly state that the omission of the namespace was intentional and maybe provide enough information to get them to some of the discussion that's happened here or elsewhere.

cjdb replied:
I've updated the description in this PR, which I _think_ becomes the main message for a squash-and-merge.

EricWF commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,100 @@
+#  error invocable_with_telemetry requires C++20
+#else
+template <class F>
+class invocable_with_telemetry {
```
Cool.

EricWF commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,100 @@
+  constexpr invocable_with_telemetry(invocable_with_telemetry&& other)
+    requires std::move_constructible<F>
+      : f_(std::move(other.f_)),
+        invocations_(assert(invocations_ != nullptr), std::exchange(other.invocations_, nullptr)),
```
This use of assert is clever, but too clever for my liking.

cjdb committed:
adds documentation and fixes telemetry move

cjdb committed:
fixes telemetry move assignment

EricWF commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,100 @@
+template <class F>
+class invocable_with_telemetry {
+public:
+  invocable_with_telemetry() = default;
```
A default constructed instance of this type cannot be copied, moved, or invoked. Was this intentional?
Given the type design that requires external memory be provided, I think we should delete the default constructor.

cjdb replied:
I'm going to delete this for now.

EricWF commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,100 @@
+#  error invocable_with_telemetry requires C++20
+#else
+template <class F>
+class invocable_with_telemetry {
```
I have a bunch of questions that stem from why/if we really need to pass around three separate addresses, one per metric, or if we can simply pass around a pointer to struct containing whatever we want?

cjdb replied:
I've added `invocable_telemetry` to structure this a bit better.

EricWF commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,100 @@
+
+  template <class... Args>
+    requires std::invocable<F&, Args...>
+  constexpr decltype(auto) operator()(Args&&... args) noexcept(std::is_nothrow_invocable_v<F&, Args...>) {
```
Does it mater if this is const-callable?

cjdb replied:
Not presently, but we might want to add this functionality in the future.

EricWF commented on libcxx/test/support/maths.h:
```diff
@@ -0,0 +1,65 @@
+
+  return result;
+}
+static_assert(factorial(0) == 1);
```
Cool beans. Love it.

EricWF commented on libcxx/test/support/maths.h:
```diff
@@ -0,0 +1,65 @@
+//===----------------------------------------------------------------------===//
+//
```
I like this header a lot. Thanks for expanding on the `gaussian_sum` idea.

EricWF reviewed (approved):
LGTM again. Thanks for spending so much time on on these tests.

cjdb committed:
fixes in_value_result

cjdb commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+} // namespace __cpo
+
+struct __fold_left {
+  template <input_iterator _Ip, sentinel_for<_Ip> _Sp, class _Tp, __indirectly_binary_left_foldable<_Tp, _Ip> _Fp>
```
There's also code that uses `_Ip` and `_Sp`. I think it would be better to have a discussion about this outside of the review about whether it's worth unifying the names.

EricWF replied:
@var-const I would be happy to review any proposal put forward for a standard naming convention in libc++. 

Different parts do things differently, and this is very much inline with existing code in libc++. I don't think its appropriate to block this change on the broader conversation about names.

var-const replied:
@EricWF This discussion has [already happened](https://discord.com/channels/636084430946959380/956590603808931870), and I can confirm that the outcome of it was to use `_Iter` and `_Sent`. Maintaining consistency is one of the purposes of doing a code review, and I don't think the fact that our code base is not perfect in this regard allows dismissing that. If you would like to reopen that discussion, you are also quite welcome to come up with a new proposal.

EricWF replied:
@var-const Sounds good! That conversation already happened, there was consensus, and now there is no longer. I would be happy to help re-establish consensus in a more inclusive forum.

cjdb replied:
I am okay with the current resolution given that a discussion has happened, and am happy to make a dedicated migration patch for all `_Ip`/`_Sp` names. This has **not** been applied in #76534: I'll make a dedicated commit for that in about an hour.

I think that the contribution guidelines should probably issue guidance on what kinds of names we expect, especially when it comes to names that we want to have some element of consistency for. Perhaps that could be the subject of the new discussion.

cjdb replied:
See #76540 for the renaming patch.

cjdb committed:
replaces rogue integers with a telemetry struct

cjdb committed:
Merge branch 'main' into fold_left

cjdb committed:
applies clang-format

cjdb committed:
adds `_LIBCPP_HIDE_FROM_ABI` to `in_value_result`

cjdb committed:
manually fixes clang-format issue

cjdb committed:
handles cxx26 action failure

cjdb committed:
fixes CI for C++20

cjdb committed:
adds tests for pointer-to-member functions and renames test functions to avoid comments

cjdb committed:
changes parameter name so it doesn't shadow

cjdb commented on libcxx/include/__algorithm/fold.h:
```diff
@@ -0,0 +1,118 @@
+    __indirectly_binary_left_foldable_impl<_Fp, _Tp, _Ip, invoke_result_t<_Fp&, _Tp, iter_reference_t<_Ip>>>;
+
+struct __fold_left_with_iter {
+  template <input_iterator _Ip, sentinel_for<_Ip> _Sp, class _Tp, __indirectly_binary_left_foldable<_Tp, _Ip> _Fp>
```
Correct, subsumption is transitive.

cjdb committed:
handle platforms that don't support certain features

EricWF reviewed (approved):
LGTM. I see @var-const has "requested changes" set, but I feel we've addressed the one unresolved comment of his.

Given the amount of work you're going to be doing in the new year adding more fold algorithms, this code is going to get a lot of love still yet.

Any concerns that haven't been satisfactorily addressed will certainly be addressed in those reviews, if not earlier.

I would like for this change to land, before we all have such good holidays that we forget all about it.


@cjdb Without further input from @var-const , I believe you should commit this change. As long as your willing to work diligently to resolve any post-commit feedback that occurs.

cjdb committed:
handle platforms that don't support certain features

cjdb commented:
Merging, with the promise to diligently resolve post-commit feedback in early January. Anything that's specific to this patch will be actioned before starting work on the remaining fold algorithms, and anything that's generalised to `std::ranges` will be actioned after finishing them.

var-const reviewed (commented):
This patch has changed substantially since I last reviewed it. I'll send more feedback in a follow-up review.

var-const commented:
> I see @var-const has "requested changes" set, but I feel we've addressed the one unresolved comment of his.

@EricWF In fact, this patch changed quite significantly since I last looked at it (based on your feedback, I believe), so I would have appreciated being able to do another round of review before this was merged.

EricWF commented on libcxx/test/std/algorithms/algorithms.results/in_value_result.pass.cpp:
```diff
@@ -0,0 +1,104 @@
+static_assert(std::is_move_constructible_v<std::ranges::in_value_result<MoveOnly, int>>);
+static_assert(std::is_move_constructible_v<std::ranges::in_value_result<int, MoveOnly>>);
+
+// should not copy constructible with move-only type
```
Attaching to this comment, because the comment about names is attached to the same line.

I think the project should document any policies we wish to enforce inside our documentation. That way we aren't depending on the visibility of conversations, and we have a place to scope our discussions of said policy. Changes to the policy can then happen in PR's changing the documentation.  Further, it better provides all contributors the ability to weigh in, hopefully at the time, but also after the fact.

To keep in with my suggestion, I will create a PR documenting our documenting policy to root this discussion under.

var-const replied:
I fully agree that adding this to a project-wide style guide would be very useful (to be clear, I don't mean to imply it has to be a comprehensive huge guide, even if it documents just a few randomly chosen aspects, it's already helpful and we can always expand later). We do have a ["Coding standards"](https://libcxx.llvm.org/Contributing.html#coding-standards) section in the `Contributing` document -- we might want to add a new page and link it from that section. Thanks for taking this on!

cjdb replied:
> I fully agree that adding this to a project-wide style guide would be very useful (to be clear, I don't mean to imply it has to be a comprehensive huge guide, even if it documents just a few randomly chosen aspects, it's already helpful and we can always expand later). We do have a ["Coding standards"](https://libcxx.llvm.org/Contributing.html#coding-standards) section in the `Contributing` document -- we might want to add a new page and link it from that section. Thanks for taking this on!

Re ultranit: applied in #76534.

Re second comment: Not sure what this is pertaining to?

EricWF commented on libcxx/test/std/algorithms/algorithms.results/in_value_result.pass.cpp:
```diff
@@ -0,0 +1,104 @@
+    std::ranges::in_value_result<ConvertibleFrom<int>, ConvertibleFrom<double>> res2 = res;
+    assert(res2.in.content == 10);
+    assert(res2.value.content == 0.);
+  }
```
Can we make clang-format do that for us?

cjdb replied:
I've applied this in #76534.

cjdb commented:
> > I see @var-const has "requested changes" set, but I feel we've addressed the one unresolved comment of his.
> 
> @EricWF In fact, this patch changed quite significantly since I last looked at it (based on your feedback, I believe), so I would have appreciated being able to do another round of review before this was merged.

That is fair. My understanding is that Eric was trying to balance respecting that you might have feedback with me struggling to switch off while OOO, if a PR is still in my "active context".

I ended up coming back into the office early, and prioritised applying your feedback as a result (see #76534). Please do make further commentary as necessary.

cjdb commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,89 @@
+#if TEST_STD_VER < 20
+#  error invocable_with_telemetry requires C++20
+#else
+struct invocable_telemetry {
```
Thoughts summarised in the comment below.

cjdb commented on libcxx/test/support/invocable_with_telemetry.h:
```diff
@@ -0,0 +1,89 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TEST_SUPPORT_INVOCABLE_WITH_TELEMETRY_H
```
My guess is "probably not", though we should come up with a consistent design (happy to work with you on that). I wasn't aware of those two, but the contents of `support/counting_predicates.h` weren't quite right for the application. I suggest we consolidate all of the counting invocable types into a single `counting_invocable` or `TracedInvocable` so that there's only one canonical type.

Looking at `TracedCopyMove`, it looks like it'll need a call operator added. Would you like me to add tests for explicitly counting copies and moves? It may be better to have these be separate tests for modularity.

Nothing has been applied yet, but I suspect a change to be made in the next week or so.

cjdb commented on libcxx/test/std/algorithms/algorithms.results/in_value_result.pass.cpp:
```diff
@@ -0,0 +1,104 @@
+};
+
+constexpr bool test() {
+  {
```
I lifted these from one of the other tests in this directory, so we should add the description to all of them. Applied in #76534.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/requirements.compile.pass.cpp:
```diff
@@ -0,0 +1,259 @@
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+// Checks that `std::ranges::fold_left_with_iter`'s requirements reject parameters that don't meet
```
Applied in #76534.

cjdb commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.fold/left_folds.pass.cpp:
```diff
@@ -0,0 +1,315 @@
+template <class Result, class T>
+concept is_dangling_with = std::same_as<Result, std::ranges::fold_left_with_iter_result<std::ranges::dangling, T>>;
+
+struct Long {
```
Agreed. Renamed in #76534.
